lib/HLSL/HLOperationLower.cpp

```diff
@@ -6871,34 +6861,57 @@ void GenerateStructBufSt(Value *handle, Value *bufIdx, Value *offset,
 }
 
 
-static Value* TranslateStructBufVecLd(Type* VecEltTy, unsigned ElemCount,
+static Value* TranslateStructOrRawBufVecLd(Type* VecEltTy, unsigned ElemCount,
   IRBuilder<>& Builder, Value* handle, hlsl::OP* OP, Value* status,
-  Value* bufIdx, Value* baseOffset, const DataLayout& DL, std::vector<Value*> &bufLds, bool isScalarTy) {
+  Value* bufIdx, Value* baseOffset, const DataLayout& DL,
+  std::vector<Value*> &bufLds, DXIL::ResourceKind kind, bool isScalarTy) {
+
+  DXASSERT_NOMSG(kind == DXIL::ResourceKind::StructuredBuffer ||
+    kind == DXIL::ResourceKind::RawBuffer);
+
   unsigned  EltSize = DL.getTypeAllocSize(VecEltTy);
+  // If RawBuffer load of 64-bit value, don't set alignment to 8,
+  // since buffer alignment isn't known to be anything over 4.
+  if (kind == HLResource::Kind::RawBuffer && EltSize > 4) {
+    EltSize = 4;
+  }
   Constant* alignment = OP->GetI32Const(EltSize);
-
   Value* offset = baseOffset;
-  if (baseOffset == nullptr)
-    offset = OP->GetU32Const(0);
+  Value* index = bufIdx;
+
+  if (baseOffset == nullptr) {
+    if (kind == DXIL::ResourceKind::RawBuffer) {
+      offset = UndefValue::get(Builder.getInt32Ty());
+    }
+    else {
+      offset = OP->GetU32Const(0);
+    }
+  }
 
   std::vector<Value*> elts(ElemCount);
   unsigned rest = (ElemCount % 4);
   for (unsigned i = 0; i < ElemCount-rest; i += 4) {
     Value* ResultElts[4];
-    Value* bufLd = GenerateStructBufLd(handle, bufIdx, offset, status, VecEltTy, ResultElts, OP, Builder, 4, alignment);
+    Value* bufLd = GenerateRawBufLd(handle, index, offset, status, VecEltTy, ResultElts, OP, Builder, 4, alignment);
     bufLds.emplace_back(bufLd);
     elts[i] = ResultElts[0];
     elts[i + 1] = ResultElts[1];
     elts[i + 2] = ResultElts[2];
     elts[i + 3] = ResultElts[3];
 
-    // Update offset by 4*4bytes.
-    offset = Builder.CreateAdd(offset, OP->GetU32Const(4 * EltSize));
+    if (kind == DXIL::ResourceKind::RawBuffer) {
+      // For rawbuffer, the validator expects offset to be undef.
+      // Therefore leave the offset as-is and increment the index instead.
+      index = Builder.CreateAdd(index, OP->GetU32Const(1));
+    }
+    else {
+      offset = Builder.CreateAdd(offset, OP->GetU32Const(4 * EltSize));
```

Comment:
Here's a place where EltSize is assumed to be the actual element size, but if it was changed above from 8 to 4, this will generate incorrect offsets for 64-bit types.